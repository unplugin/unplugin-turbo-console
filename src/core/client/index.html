<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Launch Editor</title>
</head>
<style>
  #error-info {
    color: red;
    font-size: 2rem;
  }

  body {
    background-color: white;
  }

  @media (prefers-color-scheme: dark) {
    body {
      background-color: black;
    }
  }
</style>
<body>
  <div id="error-info"></div>
</body>
<script>
 (async () => {
   try {
     const hash = window.location.hash
     const decodeBase64 = window.atob(hash.slice(1))
     await fetch(`/__open-in-editor?file=${encodeURIComponent(decodeBase64)}`)
     window.close()
   }
   catch (error) {
     const errorInfo = document.getElementById('error-info')
     errorInfo.innerText = String(error)
   }
 })()
</script>
</html>
