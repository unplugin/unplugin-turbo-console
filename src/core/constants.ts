export const PLUGIN_NAME = 'unplugin-turbo-console'
export const NUXT_CONFIG_KEY = 'turboConsole'

export const VirtualModules = {
  Init: '~console',
  ThemeDetect: '~console/theme-detect',
  VueDevTools: '~console/vue-devtools',
  ServerInfo: '~console/server-info',
}

export const PLUGIN_SERVER_DEFAULT_PORT = 3070
export const PLUGIN_SERVER_PORT_RANGE: [number, number] = [PLUGIN_SERVER_DEFAULT_PORT, 6000]
