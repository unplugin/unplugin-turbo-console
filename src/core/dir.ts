import { fileURLToPath } from 'node:url'
import { dirname, resolve } from 'pathe'

export const DIR_DIST = typeof __dirname !== 'undefined'
  ? __dirname
  : dirname(fileURLToPath(import.meta.url))

export const CLIENT_DIR = resolve(DIR_DIST, './client')
