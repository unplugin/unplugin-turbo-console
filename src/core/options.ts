import { env } from 'node:process'
import type { Options } from '../types'

const DETAULT_OPTIONS: Options = {
  prefix: '',
  suffix: '',
  disableLaunchEditor: false,
  disableHighlight: false,
  extendedPathFileNames: [],
  port: 3070,
}

export const BUILD_OPTIONS: Options = {
  disableLaunchEditor: true,
  disableHighlight: true,
}

export function resolveOptions(options: Options): Options {
  let resolved = {
    ...DETAULT_OPTIONS,
    ...options,
  }

  if (env.NODE_ENV === 'production') {
    resolved = {
      ...resolved,
      ...BUILD_OPTIONS,
    }
  }

  return resolved
}
