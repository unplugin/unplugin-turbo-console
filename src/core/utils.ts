import { sep } from 'pathe'
import { createFilter } from '@rollup/pluginutils'

export const filter = createFilter(
  [/\.vue$/, /\.vue\?vue/, /\.vue\?v=/, /\.ts$/, /\.tsx$/, /\.js$/, /\.jsx$/, /\.svelte$/],
  [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/, /[\\/]\.nuxt[\\/]/],
)

const commonStyle = 'padding:2px 5px; border-radius:3px 0 0 3px;margin-bottom:5px;'

const consoleStyles: Record<string, string> = {
  '.js': `${commonStyle}color: #111827; background: #F7DF1E`,
  '.jsx': `${commonStyle}color: #111827; background: #F7DF1E`,
  '.ts': `${commonStyle}color: #fff; background: #3178C6`,
  '.tsx': `${commonStyle}color: #fff; background: #3178C6`,
  '.vue': `${commonStyle}color: #fff; background: #4FC08D`,
  '.svelte': `${commonStyle}color: #fff; background: #FF3E00`,
  '.astro': `${commonStyle}color: #fff; background: #FF5D01`,
  'default': `${commonStyle}color: #111827; background: #F7DF1E`,
}

export function getConsoleStyle(fileType: string): string {
  return consoleStyles[fileType] ?? consoleStyles.default
}

export const launchEditorStyle = 'background: #00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin-bottom:5px'

type Framework = 'rollup' | 'vite' | 'webpack' | 'esbuild' | 'rspack'

export const getEnforce: Record<Framework, 'pre' | 'post'> = {
  rollup: 'post',
  vite: 'post',
  webpack: 'pre',
  esbuild: 'post',
  rspack: 'pre',
}

export function printInfo(port: number) {
  // eslint-disable-next-line no-console
  console.log('\x1B[32m%s\x1B[0m\x1B[1m%s\x1B[0m\x1B[36m%s\x1B[0m', '  âžœ', `  TurboConsole:`, ` http://localhost:${port}/intro`)
}

export function getSemanticPath(filePath: string, showSemanticPath?: boolean) {
  const arr = filePath.split(sep)
  let _basename = arr.pop()
  if (showSemanticPath) {
    let isEnd = false
    if (_basename?.toLowerCase().includes('index')) {
      while (!isEnd) {
        const name = arr.pop()
        _basename = `${name}/${_basename}`
        if (!name?.toLowerCase().includes('index') || !name)
          isEnd = true
      }
    }
  }
  return _basename
}
