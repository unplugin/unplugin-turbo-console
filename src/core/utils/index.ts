export * from './codegen'
export * from './common'
