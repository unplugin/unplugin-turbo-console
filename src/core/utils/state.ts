import type { Peer } from 'crossws'
import type { ExpressionMeta, ExpressionsMap } from '../../types'
import { signal } from 'alien-signals'
import { v4 as uuidv4 } from 'uuid'

export const filePathMapState = signal(new Map<string, string>())

export const expressionsMapState = signal(new Map<string, ExpressionsMap>())

export function addExpression(relativePath: string, expressionMeta: ExpressionMeta) {
  const currentMap = expressionsMapState()
  const expressions = currentMap.get(relativePath)?.expressions || []

  if (!expressions.some((item: ExpressionMeta) =>
    item.code === expressionMeta.code
    && item.method === expressionMeta.method
    && item.line === expressionMeta.line
    && item.column === expressionMeta.column,
  )) {
    const newMap = new Map(currentMap)
    newMap.set(relativePath, {
      id: uuidv4(),
      filePath: relativePath,
      expressions: [...expressions, expressionMeta],
    })
    expressionsMapState(newMap)
  }
}

export const serverState = signal(false)
export const peersState = signal(new Set<Peer>())
