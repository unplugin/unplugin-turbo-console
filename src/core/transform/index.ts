import type { Context } from './../../types'
import MagicString from 'magic-string'
import { parseSync } from 'oxc-parser'
import { genConsoleString, getCompiler, getLineAndColumn, isConsoleExpression, isPluginDisable } from '../utils'
import { walk } from '../utils/walker'
import { compilers } from './compilers'

export async function transform(context: Context) {
  const { code, id, options } = context
  const magicString = new MagicString(code)

  const compiler = await getCompiler(id)

  if (!compiler) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  const compileResult = await compilers[compiler](context)

  const oxcParsedResult = parseSync(id, compileResult.script, {
    lang: (compileResult.lang || 'js') as 'js' | 'jsx' | 'ts' | 'tsx' | undefined,
    sourceType: 'module',
  })

  const { program, comments = [] } = oxcParsedResult

  if (isPluginDisable({
    comments,
    originalLine: 1,
    id,
    type: 'top-file',
    compiler,
    script: compileResult.script,
  })) {
    return {
      code: magicString.toString(),
      map: magicString.generateMap({
        source: id,
        file: id,
        includeContent: true,
        hires: true,
      }),
    }
  }

  await walk(program, {
    enter(node) {
      if (isConsoleExpression(node)) {
        const originalExpression = compileResult.script.slice(node.start, node.end)

        if (originalExpression.includes('%c'))
          return false

        const { line, column } = getLineAndColumn(compileResult.script, node.start)

        const originalLine = line + compileResult.line
        const originalColumn = column

        if (isPluginDisable({
          comments,
          originalLine: line,
          id,
          type: 'inline-file',
          compiler,
          script: compileResult.script,
        })) {
          return false
        }

        const consoleMethod = (node as any).callee.property.name
        const args = (node as any).arguments

        const argsStart = args[0].start! + compileResult.offset
        const argsEnd = args[args.length - 1].end! + compileResult.offset
        const argType = args[0].type

        const argsName = magicString.slice(argsStart, argsEnd)
          .toString()
          .replace(/`/g, '')
          .replace(/\n/g, '')
          .replace(/"/g, '')

        const { consoleString, _suffix } = genConsoleString({
          options,
          consoleMethod,
          originalLine,
          originalColumn,
          argType,
          argsName,
          id,
        })

        if (consoleString)
          magicString.appendLeft(argsStart, consoleString)
        if (_suffix)
          magicString.appendRight(argsEnd, `,"${_suffix}"`)
      }
    },
  })

  return {
    code: magicString.toString(),
    map: magicString.generateMap({
      source: id,
      file: id,
      includeContent: true,
      hires: true,
    }),
  }
}
