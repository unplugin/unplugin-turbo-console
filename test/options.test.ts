import { join } from 'pathe'
import { describe, expect, it, vi } from 'vitest'
import { resolveOptions } from '../src/core/options'
import { transform } from '../src/core/transform'
import { filePathMapState } from '../src/core/utils/state'
import { DISABLE_ALL, DISABLE_HIGHLIGHT, DISABLE_LAUNCH_EDITOR, EMPTY, EXTENDED_PATH, WITH_PREFIX } from './fixtures/option'

vi.mock('node:process', () => {
  return {
    cwd: vi.fn(() => join('/', 'mock', 'path')),
    env: {
      NODE_ENV: 'development',
    },
  }
})

const mockFilePathMap = new Map()
mockFilePathMap.set('../../home/runner/main.js', 'fgsss')
mockFilePathMap.set('../../home/runner/index.js', 'sfgha')

filePathMapState(mockFilePathMap)

describe('transform options', () => {
  it('empty option', async () => {
    EMPTY.options = resolveOptions(EMPTY.options)

    expect(
      await transform(
        EMPTY,
      ),
    ).toMatchSnapshot()
  })

  it('with prefix suffix', async () => {
    WITH_PREFIX.options = resolveOptions(WITH_PREFIX.options)
    expect(
      await transform(
        WITH_PREFIX,
      ),
    ).toMatchSnapshot()
  })

  it('disable launch editor', async () => {
    DISABLE_LAUNCH_EDITOR.options = resolveOptions(DISABLE_LAUNCH_EDITOR.options)
    expect(
      await transform(
        DISABLE_LAUNCH_EDITOR,
      ),
    ).toMatchSnapshot()
  })

  it('disable highlight', async () => {
    DISABLE_HIGHLIGHT.options = resolveOptions(DISABLE_HIGHLIGHT.options)
    expect(
      await transform(
        DISABLE_HIGHLIGHT,
      ),
    ).toMatchSnapshot()
  })

  it('disable all', async () => {
    DISABLE_ALL.options = resolveOptions(DISABLE_ALL.options)
    expect(
      await transform(
        DISABLE_ALL,
      ),
    ).toMatchSnapshot()
  })

  it('extended path option', async () => {
    EXTENDED_PATH.options = resolveOptions(EXTENDED_PATH.options)
    expect(
      await transform(
        EXTENDED_PATH,
      ),
    ).toMatchSnapshot()
  })
})

describe('resolve options', () => {
  it('empty', () => {
    expect(resolveOptions({})).toMatchInlineSnapshot(`
      {
        "disableHighlight": false,
        "disableLaunchEditor": false,
        "disablePassLogs": false,
        "extendedPathFileNames": [],
        "port": 3070,
        "prefix": "",
        "silent": false,
        "suffix": "",
      }
    `)
  })
})
