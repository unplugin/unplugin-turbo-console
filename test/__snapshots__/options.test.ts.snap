// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform options > disable all 1`] = `
{
  "code": "console.log('hello javascript')",
  "map": SourceMap {
    "file": "main.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;

exports[`transform options > disable highlight 1`] = `
{
  "code": "console.log("%c🔦 http://localhost:3070#fgsss,1,1",\`background:#00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin:0 0 5px 0;\`,"\\n",'hello javascript')",
  "map": SourceMap {
    "file": "main.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,gIAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;

exports[`transform options > disable launch editor 1`] = `
{
  "code": "console.log("%c🚀 main.js·1",\`padding:2px 5px;border-radius:3px 0 0 3px;margin:0 0 5px 0;color:#fff;background:\${globalThis._UTC_DETECT_DARK && globalThis._UTC_DETECT_DARK() ? '#F7DF1E90;' : '#F7DF1E;'}\`,"\\n",'hello javascript')",
  "map": SourceMap {
    "file": "main.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,sMAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;

exports[`transform options > empty option 1`] = `
{
  "code": "console.log("%c🚀 main.js·1%c🔦 http://localhost:3070#fgsss,1,1",\`padding:2px 5px;border-radius:3px 0 0 3px;margin:0 0 5px 0;color:#fff;background:\${globalThis._UTC_DETECT_DARK && globalThis._UTC_DETECT_DARK() ? '#F7DF1E90;' : '#F7DF1E;'}\`,\`background:#00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin:0 0 5px 0;\`,"\\n",'hello javascript')",
  "map": SourceMap {
    "file": "main.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,6TAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;

exports[`transform options > extended path option 1`] = `
{
  "code": "console.log("%c🚀 runner/index.js·1%c🔦 http://localhost:3070#sfgha,1,1",\`padding:2px 5px;border-radius:3px 0 0 3px;margin:0 0 5px 0;color:#fff;background:\${globalThis._UTC_DETECT_DARK && globalThis._UTC_DETECT_DARK() ? '#F7DF1E90;' : '#F7DF1E;'}\`,\`background:#00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin:0 0 5px 0;\`,"\\n",'hello javascript')",
  "map": SourceMap {
    "file": "index.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,qUAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "index.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;

exports[`transform options > with prefix suffix 1`] = `
{
  "code": "console.log("-------\\n%c🚀 main.js·1%c🔦 http://localhost:3070#fgsss,1,1",\`padding:2px 5px;border-radius:3px 0 0 3px;margin:0 0 5px 0;color:#fff;background:\${globalThis._UTC_DETECT_DARK && globalThis._UTC_DETECT_DARK() ? '#F7DF1E90;' : '#F7DF1E;'}\`,\`background:#00DC8250;padding:2px 5px;border-radius:0 3px 3px 0;margin:0 0 5px 0;\`,"\\n",'hello javascript',"\\n-------")",
  "map": SourceMap {
    "file": "main.js",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,sUAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,aAAC",
    "names": [],
    "sources": [
      "main.js",
    ],
    "sourcesContent": [
      "console.log('hello javascript')",
    ],
    "version": 3,
  },
}
`;
